### PR TITLE
Add  helper methods to build http client with JKS/PKCS12 certificates

### PR DESCRIPTION
**Pull request by dormanze**
Users who use the minio client for the first time do not know how to write the SSL client code.
I suggest adding an TSL example.

### CODE REVIEW

File: examples/MinioClientWithTLS.java
Comment (balamurugana) [inline examples/MinioClientWithTLS.java:1]: This example is not needed. Refer https://minio-java.min.io/io/minio/S3Base.html#ignoreCertCheck-- and https://github.com/minio/minio-java/blob/master/api/src/main/java/io/minio/http/HttpUtils.java

**Review by balamurugana — CHANGES_REQUESTED** [review]

File: examples/MinioClientWithTLS.java
Comment (dormanze) [reply examples/MinioClientWithTLS.java:1]: Thank you for your reply.
It's nice that [HttpUtils.java](https://github.com/minio/minio-java/blob/master/api/src/main/java/io/minio/http/HttpUtils.java) provides default certificate support, but I think we should provide java-friendly jks or p12-type certificate support, and I'll be happy to contribute this part of the code if you agree.

Comment (balamurugana) [reply examples/MinioClientWithTLS.java:1]: Feel free to send a PR adding the support to `HttpUtils.java`

Comment (dormanze) [reply examples/MinioClientWithTLS.java:1]: I have added support for jks and p12 certificates in addition to the original default support for pem certificates. Please review my pr.

File: api/src/main/java/io/minio/http/HttpUtils.java [hunk 1]
```diff
@@ -126,12 +128,80 @@ public static String getHostHeader(HttpUrl url) {
...
+            .build();
+  }
+
+  /**
+   * Enable defalut external certificates
+   *
+   * @param httpClient Defalut httpClient
+   * @return OkHttpClient
+   */
+  public static OkHttpClient enableExternalCertificates(OkHttpClient httpClient)
```
Comment (balamurugana) [inline api/src/main/java/io/minio/http/HttpUtils.java:182]: We do not need this wrapper. Users know what public methods to call.

**Review by balamurugana — CHANGES_REQUESTED** [review]

File: api/src/main/java/io/minio/http/HttpUtils.java [hunk 2]
```diff
@@ -126,12 +128,80 @@ public static String getHostHeader(HttpUrl url) {
...
+   *
+   * @param truststorePath Your truststore JKS file path
+   * @param keystorePath Your keystore JKS or p12 file path
+   * @return OkHttpClient
+   */
+  public static OkHttpClient enableJKSOrP12Certificates(OkHttpClient httpClient, String truststorePath, String keystorePath)
+          throws GeneralSecurityException, IOException {
+    String keyStorePassword = System.getenv("SSL_CERT_KEYSTORE_PASSWORD");
+    String trustStorePassword = System.getenv("SSL_CERT_TRUSTSTORE_PASSWORD");
+    String keystoreType = System.getenv("SSL_CERT_KEYSTORE_TYPE");
```
[the remark this answers is not in the document]
Comment (dormanze) [reply api/src/main/java/io/minio/http/HttpUtils.java]: It's good for safety.

File: api/src/main/java/io/minio/http/HttpUtils.java [hunk 3]
```diff
@@ -126,6 +128,102 @@ public static String getHostHeader(HttpUrl url) {
...
+    /**
+     * The client uses a certificate of the p12 type.
+     *
+     * @param truststorePath Trust store file path
+     * @param keystorePath Key store file path
+     * @param trustStorePassword Trust store password
+     * @param keyStorePassword Key store password
+     * @return OkHttpClient
+     */
+  public static OkHttpClient newHttpClientByP12Cert(String truststorePath,
```
Comment (balamurugana) [inline api/src/main/java/io/minio/http/HttpUtils.java:206]: We don't want to create new http clients. Instead we would need to enable external certs. Over all this PR should be like below
```java
  private static OkHttpClient enableJKSPKCS12Certificates(
      OkHttpClient httpClient,
      String trustStorePath,
      String trustStorePassword,
      String keyStorePath,
      String keyStorePassword,
      boolean pkcs12)
      throws GeneralSecurityException, IOException {
    if (trustStorePath == null || trustStorePath.isEmpty()) {
      throw new IllegalArgumentException("trust store path must be provided");
    }
    if (trustStorePassword == null) {
      throw new IllegalArgumentException("trust store password must be provided");
    }
    if (keyStorePath == null || keyStorePath.isEmpty()) {
      throw new IllegalArgumentException("key store path must be provided");
    }
    if (keyStorePassword == null) {
      throw new IllegalArgumentException("key store password must be provided");
    }

    SSLContext sslContext = SSLContext.getInstance("TLS");
    KeyStore trustStore = KeyStore.getInstance("JKS");
    KeyStore keyStore = KeyStore.getInstance(pkcs12 ? "PKCS12" : "JKS");
    try (FileInputStream trustInput = new FileInputStream(truststorePath);
        FileInputStream keyInput = new FileInputStream(keystorePath); ) {
      trustStore.load(trustInput, trustStorePassword.toCharArray());
      keyStore.load(keyInput, keyStorePassword.toCharArray());
    }
    TrustManagerFactory trustManagerFactory =
        TrustManagerFactory.getInstance(TrustManagerFactory.getDefaultAlgorithm());
    trustManagerFactory.init(trustStore);

    KeyManagerFactory keyManagerFactory =
        KeyManagerFactory.getInstance(KeyManagerFactory.getDefaultAlgorithm());
    keyManagerFactory.init(keyStore, keyStorePassword.toCharArray());

    sslContext.init(
        keyManagerFactory.getKeyManagers(),
        trustManagerFactory.getTrustManagers(),
        new java.security.SecureRandom());

    return httpClient
        .newBuilder()
        .sslSocketFactory(
            sslContext.getSocketFactory(),
            (X509TrustManager) trustManagerFactory.getTrustManagers()[0])
        .build();
  }

  public static OkHttpClient enableJKSCertificates(
      OkHttpClient httpClient,
      String trustStorePath,
      String trustStorePassword,
      String keyStorePath,
      String keyStorePassword)
      throws GeneralSecurityException, IOException {
    return enableJKSPKCS12Certificates(
        httpClient, trustStorePath, trustStorePassword, keyStorePath, keyStorePassword, false);
  }

  public static OkHttpClient enablePKCS12Certificates(
      OkHttpClient httpClient,
      String trustStorePath,
      String trustStorePassword,
      String keyStorePath,
      String keyStorePassword)
      throws GeneralSecurityException, IOException {
    return enableJKSPKCS12Certificates(
        httpClient, trustStorePath, trustStorePassword, keyStorePath, keyStorePassword, true);
  }
```

File: examples/MinioClientWithTLS.java
Comment (balamurugana) [inline examples/MinioClientWithTLS.java:1]: Remove this obvious example.

**Review by balamurugana — CHANGES_REQUESTED** [review]

File: api/src/main/java/io/minio/http/HttpUtils.java [hunk 4]
```diff
@@ -126,6 +128,102 @@ public static String getHostHeader(HttpUrl url) {
     return host + ":" + url.port();
   }
 
+
+  private static SSLSocketFactory getSSLSocketFactory(String keystoreType, String truststorePath,
+                                                      String keystorePath, String trustStorePassword,
+                                                      String keyStorePassword)
```
[the remark this answers is not in the document]
Comment (dormanze) [reply api/src/main/java/io/minio/http/HttpUtils.java]: Verify that the certificate file path is required.

File: examples/MinioClientWithTLS.java
Comment (dormanze) [reply examples/MinioClientWithTLS.java:1]: ok

**Comment by balamurugana** [comment]
@dormanze Please fix CI failures

**Comment by dormanze** [comment]
> @dormanze Please fix CI failures

Copy that, fixed CI.

**Review by balamurugana — APPROVED** [review]

**Comment by dormanze** [comment]
@harshavardhana Please approve my pr.

**Review by harshavardhana — APPROVED** [review]

**Pull request merged**